src/uds/mod.rs

```diff
@@ -35,407 +32,116 @@ pub use read_dtc_information::*;
-            .field("Args", &self.get_payload())
-            .field("response_required", &self.response_required)
-            .finish()
+impl From<u8> for UDSErrorWrapper {
+    fn from(value: u8) -> Self {
+        if let Ok(e) = UdsError::try_from(value) {
+            Self(UDSErrorTy::Error(e))
+        } else {
+            Self(UDSErrorTy::Unknown(value))
+        }
     }
 }
 
-impl UdsCmd {
-    /// Creates a new UDS Payload
-    pub fn new(sid: UdsCommand, args: &[u8], need_response: bool) -> Self {
-        let mut b: Vec<u8> = Vec::with_capacity(args.len() + 1);
-        b.push(sid.into());
-        b.extend_from_slice(args);
-        Self {
-            bytes: b,
-            response_required: need_response,
+impl EcuNRC for UDSErrorWrapper {
+    fn desc(&self) -> String {
+        match self.0 {
+            UDSErrorTy::Error(e) =>  format!("{:?}", e),
+            UDSErrorTy::Unknown(b) => format!("Unknown error code 0x{:02X?}", b),
         }
     }
 
-    pub(crate) fn from_raw(r: &[u8], response_required: bool) -> Self {
-        Self {
-            bytes: r.to_vec(),
-            response_required,
+    fn is_ecu_busy(&self) -> bool {
+        if let UDSErrorTy::Error(e) = self.0 {
+            e == UdsError::RequestCorrectlyReceivedResponsePending
+        } else {
+            false
         }
     }
 
-    /// Returns the UDS Service ID of the command
-    pub fn get_uds_sid(&self) -> UdsCommandByte {
-        self.bytes[0].into()
-    }
-}
-
-impl BaseServerPayload for UdsCmd {
-    fn get_payload(&self) -> &[u8] {
-        &self.bytes[1..]
-    }
-
-    fn get_sid_byte(&self) -> u8 {
-        self.bytes[0]
-    }
-
-    fn to_bytes(&self) -> &[u8] {
-        &self.bytes
+    fn is_wrong_diag_mode(&self) -> bool {
+        if let UDSErrorTy::Error(e) = self.0 {
+            e == UdsError::ServiceNotSupportedInActiveSession
+        } else {
+            false
+        }
     }
 
-    fn requires_response(&self) -> bool {
-        self.response_required
+    fn is_repeat_request(&self) -> bool {
+        if let UDSErrorTy::Error(e) = self.0 {
+            e == UdsError::BusyRepeatRequest
+        } else {
+            false
+        }
     }
 }
 
-/// Base handler for UDS
-#[derive(Debug, Copy, Clone)]
-pub struct UdsVoidHandler;
-
-impl ServerEventHandler<UDSSessionType> for UdsVoidHandler {
-    #[inline(always)]
-    fn on_event(&mut self, _e: ServerEvent<UDSSessionType>) {}
+pub struct UDSProtocol{
+    session_modes: HashMap<u8, DiagSessionMode>
 }
 
-#[derive(Debug)]
-/// UDS Diagnostic server
-pub struct UdsDiagnosticServer {
-    server_running: Arc<AtomicBool>,
-    settings: Arc<RwLock<UdsServerOptions>>,
-    tx: mpsc::Sender<UdsCmd>,
-    rx: mpsc::Receiver<DiagServerResult<Vec<u8>>>,
-    repeat_count: u32,
-    repeat_interval: std::time::Duration,
-    dtc_format: Option<DTCFormatType>, // Used as a cache
+impl UDSProtocol {
+    pub fn new() -> Self {
+        let mut session_modes = HashMap::new();
```

Comment:
This seems to duplicate `UdsSessionType` content, except for the `tp_require` value (which seems to be equivalent to `if session != Default` condition)

Reply:
this offers more flexibility of registering custom session types. Since the UDSProtocol defines the protocol behavior, it makes sense to 'copy' it here, along with the description of the session modes.